File: 알고리즘/연결리스트/윤병서/boj23309.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    ios::sync_with_stdio(0);
	cin.tie(0);
	int n, m; cin >> n >> m;
	int arr[n + 1];
	for(int i = 0 ; i < n ; i ++){
		cin >> arr[i];
	}
	vector<pair<int,int>> v(1000001);
	
	for(int i = 0 ; i < n ; i ++){
		if(i == 0){
			v[arr[0]] = {arr[n - 1], arr[1]};
		}
		else if(i == n - 1){
			v[arr[n - 1]] = {arr[n - 2], arr[0]};
		}
		else{
			v[arr[i]] = {arr[i - 1], arr[i + 1]};
		}
	}
	
	
	for(int i = 0 ; i < m ; i ++){
		string a; cin >> a;
		if(a == "BN"){
			int cur, nxt;
			cin >> cur >> nxt;
			cout << v[cur].second << '\n';
			v[v[cur].second] = {nxt, v[v[cur].second].second};
			v[nxt] = {cur, v[cur].second};
			v[cur] = {v[cur].first, nxt};
		}
		else if(a == "BP"){
			int cur, prev;
			cin >> cur >> prev;
			cout << v[cur].first << '\n';
			v[v[cur].first] = {v[v[cur].first].first, prev};
			v[prev] = {v[cur].first, cur};
			v[cur] = {prev, v[cur].second};
		}
		else if(a == "CN"){
			int cur;
			cin >> cur;
			cout << v[cur].second << '\n';
			int nxt_node = v[cur].second;
			v[v[nxt_node].second] = {cur, v[v[nxt_node].second].second};
			v[cur] = {v[cur].first, v[nxt_node].second};
			v[nxt_node] = {0, 0};
		}
		else{
			int cur;
			cin >> cur;
			cout << v[cur].first << '\n';
			int prev_node = v[cur].first;
			v[v[prev_node].first] = {v[v[prev_node].first].first, cur};
			v[cur] = {v[prev_node].first, v[cur].second};
			v[prev_node] = {0, 0};
		}
	}
}

File: 알고리즘/연결리스트/윤병서/boj23309.java
package ssafy;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj23309 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] arr = new int[n];
        StringTokenizer arrToken = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(arrToken.nextToken());
        }

        Pair[] v = new Pair[1000001];

        for (int i = 0; i < n; i++) {
            if (i == 0) {
                v[arr[0]] = new Pair(arr[n - 1], arr[1]);
            } else if (i == n - 1) {
                v[arr[n - 1]] = new Pair(arr[n - 2], arr[0]);
            } else {
                v[arr[i]] = new Pair(arr[i - 1], arr[i + 1]);
            }
        }

        for (int i = 0; i < m; i++) {
        	st = new StringTokenizer(br.readLine());
            String a = st.nextToken();
            if (a.equals("BN")) {
                int cur = Integer.parseInt(st.nextToken());
                int nxt = Integer.parseInt(st.nextToken());
                System.out.println(v[cur].second);
                v[v[cur].second] = new Pair(nxt, v[v[cur].second].second);
                v[nxt] = new Pair(cur, v[cur].second);
                v[cur] = new Pair(v[cur].first, nxt);
            } else if (a.equals("BP")) {
                int cur = Integer.parseInt(st.nextToken());
                int prev = Integer.parseInt(st.nextToken());
                System.out.println(v[cur].first);
                v[v[cur].first] = new Pair(v[v[cur].first].first, prev);
                v[prev] = new Pair(v[cur].first, cur);
                v[cur] = new Pair(prev, v[cur].second);
            } else if (a.equals("CN")) {
                int cur = Integer.parseInt(st.nextToken());
                System.out.println(v[cur].second);
                int nxtNode = v[cur].second;
                v[v[nxtNode].second] = new Pair(cur, v[v[nxtNode].second].second);
                v[cur] = new Pair(v[cur].first, v[nxtNode].second);
                v[nxtNode] = new Pair(0, 0);
            } else {
                int cur = Integer.parseInt(st.nextToken());
                System.out.println(v[cur].first);
                int prevNode = v[cur].first;
                v[v[prevNode].first] = new Pair(v[v[prevNode].first].first, cur);
                v[cur] = new Pair(v[prevNode].first, v[cur].second);
                v[prevNode] = new Pair(0, 0);
            }
        }
    }

    static class Pair {
        int first;
        int second;

        public Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
}
